### ISSUE TITLE
Add new column to Ticket

### ISSUE THREAD

**Issue by tomtobac** [NullDesk/TicketDesk#125]
[**TicketDesk 2.1**]

- Visual Studio 2015
- .NET Framework 6.1 SDK

Hello!
I am trying to add a new column "Organizational Unit" at Ticket and add a new filter. 
I think I could add the new filter but I can not figure out how to add/migrate the ticket column.

I was looking for some class (like Ticket.cs) and make an migration update (I tought it has to be like that, i am new (very new, but i tried like a week on this) on asp.net) but I can not find out. I have find a class called TicketPartial.cs which is [MetadataType(typeof(TicketMeta))] which (TicketMeta.cs) 
is auto-generated from a template and don't know where is the template or idea where to find out. 
Could I just edit TicketMeta.cs and add my new property?

Could someone help me and guide me how to do it without delete the current tickets?! I would appreciate a lot.

Attached: I did the function to get the *Organizational Unit* from samAccountName @ Domain\Model\AdDataRepository.cs. Is it ok there? Could be better? Thanks a lot!

```
internal string GetOrganizationalUnitUser(string samAccountName)
        {
            using
           (
               PrincipalContext ctx = new PrincipalContext(ContextType.Domain,
                                       GetActiveDirectoryDomain(),
                                       GetActiveDirectoryUser(),
                                       GetActiveDirectoryUserPassword())
           )
            {

// Our company we have AD_user which works in OU called "Users" and it belongs to another OU which 
// has a representative name which I want
// Example: { Company1: {Users: "myUser1"}} -- "myUser1" it belongs to "Company1"

                if(samAccountName == null)
                {
                    return "Mail";
                }

// We return "Mail" because we have Tickets as well from eMail and maybe they are not users in 
// AD so we group them in "Mail"

                var user = Principal.FindByIdentity(ctx, samAccountName);

                if (user != null) 
                {
                    using (DirectoryEntry deUser = user.GetUnderlyingObject() as DirectoryEntry)
                    {
                        using (DirectoryEntry deUserContainer = deUser.Parent)
                        {
                            if (deUserContainer != null) 
                            {
                                using (DirectoryEntry deUserContainerUpper = deUserContainer.Parent)
                                {
                                    if (deUserContainerUpper != null)
                                    {
                                        return deUserContainerUpper.Properties["Name"].Value.ToString();
                                    }
                                    else
                                    {
                                        return "Mail";
                                    }
                                }
                            }
                            else
                            {
                                return "Mail";
                            }
                        }
                    }
                }
                else
                {
                    return "Mail";
                }
            }
               
        }

```

**Comment by StephenRedd** [comment]
Ticketdesk 2.1 is an older application, and doesn't use migrations or EF code-first techniques; it uses the old EF database first model diagrams. 

The *Meta classes are a special "trick" from the EF 4 days. Since the actual entity classes are generated from the edmx diagram, there isn't any way to add custom attributes that EF wasn't originally designed for. 

The meta classes were a trick that allowed you to specify a surrogate class that had the same property names as the real model class, then you can add custom attributes of your own to it. The meta class itself doesn't actually get used at runtime, it's just there so EF can use reflection to grab additional attributes and map them over to the properties of real model class.

Meta classes are used here so we can add validation attributes from data annotations, which the UI uses for field validation. 

There is also a partial class for some entities, TicketPartial.cs for example. Those are useful if you want to add custom methods or properties that are NOT related to the actual DB operations... stuff that's only needed in code at runtime.

Anyway, to add a new column:

1.  Add the column to the DB
  - No migrations here, use SQL Managment Studio or the SQL data tools build into VS 
2. Open TicketDeskEntities.edmx and add the column to the correct entity
 - Pay attention to the properties and make sure you are careful to set the datatype and such correctly
3. Open the *Meta.cs file for the entity you modified and manually add a property with the same name
 - Configure any data annotations attributes you want for the new property 

If you want to be a competitionist, TD does have a database schema manager (my poor-man's attempt at migrations from back before EF had migrations). 

This is in DatabaseSchemaManagerRepository.cs; you could modify this to add the SQL commands needed to add your custom field to the DB, and increment the DB version. That's a lot of work though if you don't plan to redistribute the code.

**Comment by StephenRedd** [comment]
I wanted to also mention; if you open the edmx diagram in VS, there is an update model from database option. You might be able to use this, but I'm not certain if it would try to make other changes to other entities that you might not want it to make --if you use this tool, I'd recommend doing a compare of the version in source and the changes after the tool so you can be sure it didn't try to do something stupid.

**Comment by tomtobac** [comment]
Thanks for great answer @StephenRedd but yesterday I could do it!
Anyways here is how I did, very similar your second answer to save actual Tickets alive!

1. updated model from database (I added the new column)
2. In propierties of model changed "Generation Code Strategy" for "Object context heredity" thats what i was stucked because It didn't generate the code.
3. Modify the constructor of Ticket to add the new property.
4. Edit "CreateNewTicket" function the new property and set the value (OrganizativeUnit, in my case)
5. Add filters and views.

Done! Actually, I am working like a Intern and they assigned my this work, with you code, Internet and help of my teammates i could learn a lot of MVC & asp.net/C#. 

**Thanks a lot** for the help and your work!
